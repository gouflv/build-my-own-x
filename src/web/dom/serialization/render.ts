export const render = vdom => {
  if (typeof vdom === 'string') return document.createTextNode(vdom)

  const el = document.createElement(vdom.type)

  const { children, ...restProps } = vdom.props || {}

  if (restProps)
    for (let k in restProps) {
      el.setAttribute(k === 'className' ? 'class' : k, restProps[k])
    }

  if (children) {
    if (typeof children === 'string') {
      el.appendChild(document.createTextNode(children))
    } else {
      children.forEach(child => {
        el.appendChild(render(child))
      })
    }
  }

  return el
}
