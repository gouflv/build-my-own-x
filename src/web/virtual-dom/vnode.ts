export class Element<T extends keyof HTMLElementTagNameMap> {
  constructor(
    public tagName: T,
    public attr: Partial<HTMLElementTagNameMap[T]> | null | undefined,
    public children: (Element<any> | string)[]
  ) {}

  render() {
    const node = document.createElement(this.tagName)
    this.setNodeAttr(node)

    if (this.children.length) {
      this.children.forEach(child => {
        if (child instanceof Element) {
          node.appendChild(child.render())
        }
        if (typeof child === 'string') {
          node.appendChild(document.createTextNode(child))
        }
      })
    }

    return node
  }

  private setNodeAttr(node: HTMLElementTagNameMap[T]) {
    for (let key in this.attr) {
      if (!this.attr.hasOwnProperty(key)) {
        return
      }
      if (key === 'style') {
        node.style.cssText = this.attr[key] as any
      } else {
        node.setAttribute(key, this.attr[key] as any)
      }
    }
  }
}

export const createElement = <T extends keyof HTMLElementTagNameMap>(
  tagName: Element<T>['tagName'],
  attr?: Element<T>['attr'],
  children?: Element<T>['children']
) => {
  return new Element(tagName, attr || {}, children || [])
}
