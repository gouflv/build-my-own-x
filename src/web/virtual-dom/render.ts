import { Element } from './vnode'

export const render = (el: Element<any>, container: HTMLElement) => {
  container.appendChild(el.render())
}
