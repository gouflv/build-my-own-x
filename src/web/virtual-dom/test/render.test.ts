import { render } from '../render'
import { createElement as h } from '../vnode'

describe('Test render', () => {
  const ct = document.createElement('div')

  it('simple', () => {
    const el = h('div', { className: 'root' }, [h('span', null, ['Hi'])])
    render(el, ct)
    expect(ct).toMatchSnapshot()
  })
})
