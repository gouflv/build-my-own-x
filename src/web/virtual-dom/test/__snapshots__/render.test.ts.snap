// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render simple 1`] = `
<div>
  <div
    classname="root"
  >
    <span>
      Hi
    </span>
  </div>
</div>
`;
